lib/new-admin/graphql/modules/authentication.js

```diff
@@ -0,0 +1,283 @@
+      if (!isCodeValid) throw new authErrors.InvalidTwoFactorError()
+
+      return users.getUserById(user.id)
+    })
+    .then(user => {
+      const finalUser = { id: user.id, username: user.username, role: user.role }
+      context.req.session.user = finalUser
+      if (rememberMe) context.req.session.cookie.maxAge = REMEMBER_ME_AGE
+
+      return users.save2FASecret(user.id, secret)
+    })
+    .then(() => true)
+}
+
+const changeUserRole = (code, id, newRole, context) => {
+  const action = (id, newRole) => users.changeUserRole(id, newRole)
+
+  if (!code) {
+    return action(id, newRole)
+  }
+
+  return confirm2FA(code, context)
+    .then(() => action(id, newRole))
+}
+
+const enableUser = (code, id, context) => {
+  const action = id => users.enableUser(id)
+
+  if (!code) {
+    return action(id)
+  }
+
+  return confirm2FA(code, context)
+    .then(() => action(id))
+}
+
+const disableUser = (code, id, context) => {
+  const action = id => users.disableUser(id)
+
+  if (!code) {
+    return action(id)
+  }
+
+  return confirm2FA(code, context)
+    .then(() => action(id))
+}
+
+const createResetPasswordToken = (code, userID, context) => {
+  const action = userID => {
+    return users.getUserById(userID)
+      .then(user => {
+        if (!user) throw new authErrors.InvalidCredentialsError()
+        return users.createAuthToken(user.id, 'reset_password')
+      })
+      .catch(err => console.error(err))
+  }
+
+  if (!code) {
+    return action(userID)
+  }
+
+  return confirm2FA(code, context)
+    .then(() => action(userID))
+}
+
+const createReset2FAToken = (code, userID, context) => {
+  const action = userID => {
+    return users.getUserById(userID)
+      .then(user => {
+        if (!user) throw new authErrors.InvalidCredentialsError()
+        return users.createAuthToken(user.id, 'reset_twofa')
+      })
+      .catch(err => console.error(err))
+  }
+
+  if (!code) {
+    return action(userID)
+  }
+
+  return confirm2FA(code, context)
+    .then(() => action(userID))
+}
+
+const createRegisterToken = (username, role) => {
+  return users.getUserByUsername(username)
+    .then(user => {
+      if (user) throw new authErrors.UserAlreadyExistsError()
+
+      return users.createUserRegistrationToken(username, role)
+    })
+    .catch(err => console.error(err))
+}
+
+const register = (token, username, password, role) => {
+  return users.getUserByUsername(username)
+    .then(user => {
+      if (user) throw new authErrors.UserAlreadyExistsError()
+      return users.register(token, username, password, role).then(() => true)
+    })
+    .catch(err => console.error(err))
+}
+
+const resetPassword = (token, userID, newPassword, context) => {
+  return users.getUserById(userID)
+    .then(user => {
+      if (!user) throw new authErrors.InvalidCredentialsError()
+      if (context.req.session.user && user.id === context.req.session.user.id) context.req.session.destroy()
```

Comment:
Instead of using context manually in a lot of places you can create functions that does what you want using the context. If we do that we will have easier to follow code such as:

```js
       if (!user) throw new authErrors.InvalidCredentialsError()
       destroySessionIfSameUser(context, user)
       return....
```